core/kotlinx-coroutines-core/src/test_/TestCoroutineExceptionHandler.kt

```diff
@@ -0,0 +1,73 @@
+package kotlinx.coroutines.test
+
+import kotlinx.coroutines.CoroutineExceptionHandler
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Access uncaught coroutines exceptions captured during test execution.
+ *
+ * Note, tests executed via [runBlockingTest] or [TestCoroutineScope.runBlocking] will not trigger uncaught exception
+ * handling and should use [Deferred.await] or [Job.getCancellationException] to test exceptions.
+ */
+interface ExceptionCaptor {
+    /**
+     * List of uncaught coroutine exceptions.
+     *
+     * Tests must process these exceptions prior to [cleanupTestCoroutines] either by clearing this list with
+     * [exceptions.clear] or calling [rethrowUncaughtCoroutineException].
+     */
+    val exceptions: MutableList<Throwable>
+
+    /**
+     * Rethrow the first uncaught coroutine exception immediately.
+     *
+     * This allows tests to use their preferred exception testing techniques.
+     *
+     * If a test generates uncaught exceptions, it must call this method, or clear [exceptions] prior to calling
+     * [cleanupTestCoroutines].
+     */
+    fun rethrowUncaughtCoroutineException(): Nothing
```

Comment:
The exception checking mechanisms provided by some common unit testing libraries (kotlin.test's `assertFailWith`, JUnit's `ExpectedException` rule) will already fail tests if the code that is expected to throw does not. I would rather not have this method throw its own exception in those cases, to allow testing libraries to fail consistently as they would in other cases.

Reply:
Yea - agreed. This API is adding complexity for not much gain. I'll reduce it to the list of exceptions, and rethrow the first uncaught exception in the cleanup method.

Reply:
Test code that wants this behavior can just do

`throw exceptions.firstOrNull()`

Reply:
Awesome, that also makes it really hard to forget that there might be multiple uncaught exceptions. And it's easy to add this function back later.